File: app/src/main/java/com/druger/aboutwork/fragments/UserReviewsFragment.java
package com.druger.aboutwork.fragments;


import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.fragment.app.FragmentTransaction;
import androidx.recyclerview.widget.DefaultItemAnimator;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.arellomobile.mvp.presenter.InjectPresenter;
import com.bumptech.glide.Glide;
import com.bumptech.glide.request.RequestOptions;
import com.druger.aboutwork.R;
import com.druger.aboutwork.activities.MainActivity;
import com.druger.aboutwork.adapters.MyReviewAdapter;
import com.druger.aboutwork.interfaces.OnItemClickListener;
import com.druger.aboutwork.interfaces.view.UserReviews;
import com.druger.aboutwork.model.Review;
import com.druger.aboutwork.presenters.UserReviewsPresenter;
import com.google.firebase.storage.StorageReference;

import java.util.ArrayList;
import java.util.List;

import static com.druger.aboutwork.Const.Bundles.USER_ID;

public class UserReviewsFragment extends BaseSupportFragment implements UserReviews {

    @InjectPresenter
    UserReviewsPresenter reviewsPresenter;

    private ImageView civAvatar;
    private TextView tvName;

    private RecyclerView rvReviews;
    private MyReviewAdapter reviewAdapter;
    private List<Review> reviews = new ArrayList<>();

    public UserReviewsFragment() {
        // Required empty public constructor
    }

    public static UserReviewsFragment newInstance(String userId) {

        Bundle args = new Bundle();

        UserReviewsFragment fragment = new UserReviewsFragment();
        args.putString(USER_ID, userId);
        fragment.setArguments(args);
        return fragment;
    }


    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        rootView = inflater.inflate(R.layout.fragment_user_reviews, container, false);
        setupUI();
        setupRecycler(reviews);
        reviewsPresenter.downloadPhoto(getArguments().getString(USER_ID));
        reviewsPresenter.fetchReviews(getArguments().getString(USER_ID));
        reviewsPresenter.getUserName(getArguments().getString(USER_ID));
        setupToolbar();
        ((MainActivity) getActivity()).hideBottomNavigation();
        return rootView;
    }

    private void setupToolbar() {
        mToolbar = bindView(R.id.toolbar);
        setActionBar(mToolbar);
        getActionBar().setDisplayHomeAsUpEnabled(true);
    }

    private void setupRecycler(final List<Review> reviews) {
        LinearLayoutManager layoutManager = new LinearLayoutManager(getActivity());
        reviewAdapter = new MyReviewAdapter(reviews);
        rvReviews.setLayoutManager(layoutManager);
        rvReviews.setItemAnimator(new DefaultItemAnimator());
        rvReviews.setAdapter(reviewAdapter);

        reviewAdapter.setOnClickListener(new OnItemClickListener<Review>() {
            @Override
            public void onClick(Review review, int position) {
                SelectedReviewFragment reviewFragment = SelectedReviewFragment.newInstance(review.getFirebaseKey(), false);

                FragmentTransaction transaction = getFragmentManager().beginTransaction();
                transaction.replace(R.id.main_container, reviewFragment);
                transaction.addToBackStack(null);
                transaction.commit();
            }

            @Override
            public boolean onLongClick(int position) {
                return false;
            }
        });
    }

    private void setupUI() {
        civAvatar = bindView(R.id.ivAvatar);
        tvName = bindView(R.id.tvName);
        rvReviews = bindView(R.id.rvReviews);
    }

    @Override
    public void showPhoto(StorageReference storageRef) {
        Glide.with(getActivity())
                .load(storageRef)
                .apply(RequestOptions.circleCropTransform())
                .error(R.drawable.ic_account_circle_black)
                .into(civAvatar);
    }

    @Override
    public void notifyDataSetChanged() {
        reviewAdapter.notifyDataSetChanged();
    }

    @Override
    public void showReviews(List<Review> reviews) {
        this.reviews.clear();
        this.reviews.addAll(reviews);
        reviewAdapter.notifyDataSetChanged();

    }

    @Override
    public void showName(String name) {
        tvName.setText(name);
        getActionBar().setTitle(name);
    }

    @Override
    public void onStop() {
        super.onStop();
        reviewsPresenter.removeListeners();
    }
}

File: app/src/main/java/com/druger/aboutwork/fragments/UserReviewsFragment.kt
package com.druger.aboutwork.fragments


import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.core.view.isVisible
import androidx.fragment.app.viewModels
import androidx.recyclerview.widget.DefaultItemAnimator
import com.druger.aboutwork.Const.Bundles.USER_ID
import com.druger.aboutwork.R
import com.druger.aboutwork.activities.MainActivity
import com.druger.aboutwork.adapters.MyReviewAdapter
import com.druger.aboutwork.databinding.FragmentUserReviewsBinding
import com.druger.aboutwork.interfaces.OnItemClickListener
import com.druger.aboutwork.model.Review
import com.druger.aboutwork.viewmodels.UserReviewsViewModel
import dagger.hilt.android.AndroidEntryPoint

@AndroidEntryPoint
class UserReviewsFragment : BaseSupportFragment() {

    private val viewModel: UserReviewsViewModel by viewModels()

    private var _binding: FragmentUserReviewsBinding? = null
    private val binding get() = _binding!!

    private var reviewAdapter: MyReviewAdapter? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        observeReviewState()
        observeNameState()
    }

    private fun observeNameState() {
        viewModel.nameState.observe(this) { showName(it) }
    }

    private fun observeReviewState() {
        viewModel.updateReviewState.observe(this) { notifyDataSetChanged() }
        viewModel.reviewsState.observe(this) { showReviews(it) }
    }

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        _binding = FragmentUserReviewsBinding.inflate(inflater, container, false)
        arguments?.getString(USER_ID)?.let { viewModel.fetchReviews(it) }
        arguments?.getString(USER_ID)?.let { viewModel.getUserName(it) }
        (activity as MainActivity).hideBottomNavigation()
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        setupToolbar()
        setupRecycler()
    }

    private fun setupToolbar() {
        setActionBar(binding.toolbar.toolbar)
        actionBar?.setDisplayHomeAsUpEnabled(true)
    }

    private fun setupRecycler() {
        reviewAdapter = MyReviewAdapter()
        binding.rvReviews.itemAnimator = DefaultItemAnimator()
        binding.rvReviews.adapter = reviewAdapter

        reviewAdapter?.setOnClickListener(object : OnItemClickListener<Review> {
            override fun onClick(item: Review, position: Int) {
                item.firebaseKey?.let {
                    val reviewFragment = SelectedReviewFragment.newInstance(it, false)

                    fragmentManager?.beginTransaction()?.apply {
                        replace(R.id.main_container, reviewFragment)
                        addToBackStack(null)
                        commit()
                    }
                }
            }

            override fun onLongClick(item: Review, position: Int): Boolean {
                return false
            }
        })
    }

    private fun notifyDataSetChanged() {
        reviewAdapter?.notifyDataSetChanged()
    }

    private fun showReviews(reviews: List<Review>) {
        with(binding) {
            if (reviews.isNotEmpty()) {
                reviewAdapter?.addReviews(reviews)
                tvCountReviews.text =
                    resources.getQuantityString(R.plurals.reviews, reviews.size, reviews.size)
            } else {
                groupReviews.isVisible = false
                ltNoReviews.root.isVisible = true
                ltNoReviews.tvNoReviews.text = getString(R.string.user_no_reviews)
            }
        }
    }

    private fun showName(name: String) {
        actionBar?.title = name
    }

    override fun onStop() {
        super.onStop()
        viewModel.removeListeners()
    }

    companion object {

        fun newInstance(userId: String): UserReviewsFragment {

            val args = Bundle()

            val fragment = UserReviewsFragment()
            args.putString(USER_ID, userId)
            fragment.arguments = args
            return fragment
        }
    }
}
